File: example/java/jimmer-sql-graphql/src/main/java/org/babyfish/jimmer/sql/example/graphql/controller/BookController.java
package org.babyfish.jimmer.sql.example.graphql.controller;

import org.babyfish.jimmer.sql.JSqlClient;
import org.babyfish.jimmer.sql.ast.mutation.AffectedTable;
import org.babyfish.jimmer.sql.example.graphql.dal.BookRepository;
import org.babyfish.jimmer.sql.example.graphql.entities.*;
import org.babyfish.jimmer.sql.example.graphql.entities.input.BookInput;
import org.springframework.graphql.data.method.annotation.Argument;
import org.springframework.graphql.data.method.annotation.BatchMapping;
import org.springframework.graphql.data.method.annotation.MutationMapping;
import org.springframework.graphql.data.method.annotation.QueryMapping;
import org.springframework.lang.Nullable;
import org.springframework.stereotype.Controller;
import org.springframework.transaction.annotation.Transactional;

import java.util.Collection;
import java.util.List;
import java.util.Map;

@Controller
public class BookController {

    private final JSqlClient sqlClient;

    private final BookRepository bookRepository;

    public BookController(
            JSqlClient sqlClient,
            BookRepository bookRepository
    ) {
        this.sqlClient = sqlClient;
        this.bookRepository = bookRepository;
    }

    // --- Query ---

    @QueryMapping
    public List<Book> books(
            @Argument @Nullable String name,
            @Argument @Nullable String storeName,
            @Argument @Nullable String authorName
    ) {
        return bookRepository.find(name, storeName, authorName);
    }

    // --- Association ---

    @BatchMapping
    public Map<Book, BookStore> store(Collection<Book> books) {
        return sqlClient
                .getLoaders()
                .reference(BookProps.STORE)
                .batchLoad(books);
    }

    @BatchMapping
    public Map<Book, List<Author>> authors(List<Book> books) {
        return sqlClient
                .getLoaders()
                .list(BookProps.AUTHORS)
                .batchLoad(books);
    }

    // --- Mutation ---

    @MutationMapping
    @Transactional
    public Book saveBook(@Argument BookInput input) {
        return sqlClient.getEntities().save(input.toBook()).getModifiedEntity();
    }

    @MutationMapping
    @Transactional
    public int deleteBook(@Argument long id) {
        return sqlClient
                .getEntities()
                .delete(Book.class, id)
                .getAffectedRowCount(AffectedTable.of(Book.class));
    }
}

File: example/kotlin/jimmer-sql-graphql-kt/src/main/kotlin/org/babyfish/jimmer/example/kt/graphql/controller/BookController.kt
package org.babyfish.jimmer.example.kt.graphql.controller

import org.babyfish.jimmer.example.kt.graphql.dal.BookRepository
import org.babyfish.jimmer.example.kt.graphql.entities.Author
import org.babyfish.jimmer.example.kt.graphql.entities.Book
import org.babyfish.jimmer.example.kt.graphql.entities.BookStore
import org.babyfish.jimmer.example.kt.graphql.entities.input.BookInput
import org.babyfish.jimmer.sql.kt.KSqlClient
import org.springframework.graphql.data.method.annotation.Argument
import org.springframework.graphql.data.method.annotation.BatchMapping
import org.springframework.graphql.data.method.annotation.MutationMapping
import org.springframework.graphql.data.method.annotation.QueryMapping
import org.springframework.stereotype.Controller
import org.springframework.transaction.annotation.Transactional

@Controller
class BookController(
    private val sqlClient: KSqlClient,
    private val bookRepository: BookRepository
) {

    // --- Query ---

    @QueryMapping
    fun books(
        @Argument name: String?,
        @Argument storeName: String?,
        @Argument authorName: String?
    ): List<Book> =
        bookRepository.find(name, storeName, authorName)

    // --- Association ---

    @BatchMapping
    fun store(
        // Must use `java.util.List` because Spring-GraphQL has a bug: #454
        books: java.util.List<Book>
    ): Map<Book, BookStore> =
        sqlClient
            .loaders
            .reference(Book::store)
            .batchLoad(books)

    @BatchMapping
    fun authors(
        // Must use `java.util.List` because Spring-GraphQL has a bug: #454
        books: java.util.List<Book>
    ): Map<Book, List<Author>> =
        sqlClient
            .loaders
            .list(Book::authors)
            .batchLoad(books)

    // --- Mutation ---

    @MutationMapping
    @Transactional
    fun saveBook(@Argument input: BookInput): Book =
        sqlClient.entities.save(input.toBook()).modifiedEntity

    @MutationMapping
    @Transactional
    fun deleteBook(@Argument id: Long): Int {
        return sqlClient
            .entities
            .delete(Book::class, id)
            .affectedRowCount(Book::class)
    }
}